File: 2gift1/gift1.cc
/*
ID: cellofo1
PROG: gift1
LANG: C++
*/
#include <fstream>
#include <iostream>
#include <map>
#include<sstream>
#include<stdio.h>
#include<stdlib.h>
#include<string>
#include<vector>
using namespace std;

ifstream fin("gift1.in");
ofstream fout("gift1.out");

int main() {
    //put all the names in the map
    map<string,int> friends;
    map<string,int> ufriends;
    string nu; getline(fin,nu);
    int n = atoi(nu.c_str());
    vector<string> nombre;
    nombre.resize(n);
    nombre.insert(2,"d");
    for(int i = 0; i<n; i++)
        {
            string name; getline(fin, name);
            friends.insert(std::pair<string,int>(name,-9999));
            //nombre.push_back(name);
        }
    ufriends = friends; //ufriends is a copy of friends which will be updated later
    for(int i = 0; i<n; i++)
    {
        string name; getline(fin, name);
        string initial; getline(fin, initial, ' ');
        if(friends.at(name)== -9999)
        {
            int val = atoi(initial.c_str());
            friends.at(name)=val;
            ufriends.at(name)=val;
        }
        string div; getline(fin, div); //number of people to give money to
        int ii = atoi(initial.c_str());
        int di = atoi(div.c_str());
        float give = (float)ii/di; //amount for each person
        int giv = give; //integer division
        for(int i = 0; i<di;i++)  //distribute the monies
        {
            string asdf; getline(fin, asdf); //name of person to give to
            if(friends.at(asdf) == -9999)
            {
                friends.at(asdf) = giv;
                ufriends.at(asdf)= giv;
            }
            if(ufriends.at(asdf)==-9999)
                ufriends.at(asdf)=giv;
            else
                ufriends.at(asdf)=ufriends.at(asdf)+giv;
        }
        ufriends.at(name)=ufriends.at(name)-giv*atoi(div.c_str());
    }
    for(int j = 0; j<n;j++)
    {
//       fout<< nombre.at(j) <<" "<< ufriends.at(nombre.at(j))-friends.at(nombre.at(j))<< '\n';
    }
    return 0;
}

File: 2gift1/gift1.java
/*
ID: cellofo1
PROG: gift1
LANG: JAVA
*/
import java.util.*;
import java.io.*;

class gift1
{
    public static void main(String[] args) throws IOException
    {
        int large = -99999999;
        BufferedReader fin = new BufferedReader(new FileReader("gift1.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("gift1.out")));
        int N = Integer.parseInt(fin.readLine());
        String[] names = new String[N];
        Map<String, Integer> foo = new HashMap<String, Integer>();
        Map<String, Integer> foo1 = new HashMap<String, Integer>();
        for(int i = 0; i<N; i++)
        {
            names[i]=fin.readLine();
            foo.put(names[i],large);
            foo1.put(names[i],large);
        }
        for(int i = 0; i<N; i++)
        {
            String name = fin.readLine();
            StringTokenizer st = new StringTokenizer(fin.readLine());
            int initial = Integer.parseInt(st.nextToken());
            int div = Integer.parseInt(st.nextToken());
            if(foo.get(name)==large)
            {
                foo.put(name,initial);
                foo1.put(name,initial);
            }
            int give;
            if(div==0)
                give = 0;
            else
                give = initial/div;
            for(int k = 0; k<div; k++)
            {
                String dn = fin.readLine();
                if(foo.get(dn)==large)
                {
                    foo.put(dn,give);
                    foo1.put(dn,give);
                }
                if(foo1.get(dn)==large)
                    foo1.put(dn,give);
                else
                    foo1.put(dn,foo1.get(dn)+give);
            }
            foo1.put(name, foo1.get(name)-give*div);
        }
        for(int j = 0; j<N; j++)
        {
            int d = foo1.get(names[j])-foo.get(names[j]);
            out.println(names[j]+" "+d);
        }
        out.close();
        System.exit(0);        
    }
}
